### PR TITLE
Interpreter: cast proc call arguments to proc arg types

### LINKED ISSUE #12350: Interpreter: Reached the unreachable with athena-validator callback constraint
## Bug Report
Example validator usage:
```crystal
require "../../src/athena-validator"

class A
end

callback = AVD::Constraints::Callback::CallbackProc.new do |value, context|
  value2 = value.get(A)
  context.add_violation "different violation"
end
constraint = AVD::Constraints::Callback.new callback: callback

pp AVD.validator.validate(A.new, constraint)
```
Works fine in compiled mode but fails in interpreted mode:
```
In: Athena::Validator::Constraints::Callback::ValueContainer+#get
Node: raise "unreachable"
Reached the unreachable (Exception)
  from /home/chao/git/personal/crystal/src/compiler/crystal/interpreter/interpreter.cr:337:7 in 'interpret'
  from /home/chao/git/personal/crystal/src/compiler/crystal/interpreter/interpreter.cr:235:13 in 'interpret'
  from /home/chao/git/personal/crystal/src/compiler/crystal/interpreter/repl.cr:100:5 in 'interpret_and_exit_on_error'
  from /home/chao/git/personal/crystal/src/gc/boehm.cr:129:5 in 'run'
  from /home/chao/git/personal/crystal/src/compiler/crystal/command.cr:51:5 in '__crystal_main'
  from /home/chao/git/personal/crystal/src/crystal/main.cr:115:5 in 'main'
  from /lib/x86_64-linux-gnu/libc.so.6 in '??'
  from /lib/x86_64-linux-gnu/libc.so.6 in '__libc_start_main'
  from /home/chao/git/personal/crystal/.build/crystal in '_start'
  from ???
```
And you'll get `index out of bounds` error after changing `value2` to `value` in the example code:
```
In: Athena::Validator::Constraints::Callback::ValueContainer+#get
Node: self.is_a?(Athena::Validator::Constraints::Callback::Value(A))
Index out of bounds (IndexError)
  from /home/chao/git/personal/crystal/src/pointer.cr:516:5 in 'interpret'
  from /home/chao/git/personal/crystal/src/compiler/crystal/interpreter/interpreter.cr:235:13 in 'interpret'
  from /home/chao/git/personal/crystal/src/compiler/crystal/interpreter/repl.cr:100:5 in 'interpret_and_exit_on_error'
  from /home/chao/git/personal/crystal/src/gc/boehm.cr:129:5 in 'run'
  from /home/chao/git/personal/crystal/src/compiler/crystal/command.cr:51:5 in '__crystal_main'
  from /home/chao/git/personal/crystal/src/crystal/main.cr:115:5 in 'main'
  from /lib/x86_64-linux-gnu/libc.so.6 in '??'
  from /lib/x86_64-linux-gnu/libc.so.6 in '__libc_start_main'
  from /home/chao/git/personal/crystal/.build/crystal in '_start'
  from ???
```

Crystal version:
```
Crystal 1.6.0-dev [8cb915f95] (2022-08-01)

LLVM: 14.0.6
Default target: x86_64-pc-linux-gnu

OS: Ubuntu 22.04
```

### PR DESCRIPTION
**Pull request by asterite**
Fixes #12350

Note that for the example in #12350 to work it also needs #12374 merged.

### CODE REVIEW

**Comment by asterite** [comment]
It seems `Fiber#@proc` is of type `Proc(Nil) | Proc(Void)` 😅 . I think those two types should be unified.

I'll try to fix it here too... that shouldn't happen.

**Comment by asterite** [comment]
Actually, I'll just fix the interpreter for now.

**Comment by asterite** [comment]
I can confirm #12388 makes this work without the workaround, so we might want to merge that one first.

**Comment by asterite** [comment]
I just undid the workaround. Let's see if CI passes now :-)

File: src/compiler/crystal/interpreter/primitives.cr [hunk 1]
```diff
@@ -254,7 +254,19 @@ class Crystal::Repl::Compiler
...
       pointer_address(node: node)
     when "proc_call"
-      node.args.each { |arg| request_value(arg) }
+      proc_type = (obj.try(&.type) || scope).as(ProcInstanceType)
+
+      node.args.each_with_index do |arg, arg_index|
+        request_value(arg)
+
+        # Cast call argument to proc's type
+        # (this same logic is done in codegen/primtiives.cr)
```
Comment (Sija) [inline src/compiler/crystal/interpreter/primitives.cr:263]:
```suggestion
        # (this same logic is done in codegen/primitives.cr)
```

Comment (asterite) [reply src/compiler/crystal/interpreter/primitives.cr]: Thanks!

**Comment by asterite** [comment]
CI failed and I have a fix for this. But I found out that a lot more primitive calls are broken if they happen inside a closure, so I'm going to do a general fix for that.

**Comment by asterite** [comment]
So, the last non-merge commit is doing this. If you had code like this:

```crystal
struct Enum
  def foo
    ->{
      value
    }
  end
end

enum Color
  Red
  Green
end

p Color::Green.foo.call
```

the "scope" (the type of `self`) inside the proc literal was incorrectly `Program` instead of `Enum` (or `Color`) in this case. This affected all primitive calls that happened without a receiver in a proc literal, so it was affecting `Proc#call` and probably many others. The last commit preserves the correct scope inside proc literals.

**Review by straight-shoota — APPROVED** [review]

**Pull request merged**